temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -2094,39 +2085,68 @@ public void signalFromWorkflow(SignalExternalWorkflowExecutionCommandAttributes
   @Override
   public UpdateWorkflowExecutionResponse updateWorkflowExecution(
       UpdateWorkflowExecutionRequest request, Deadline deadline) {
+    if (request
+            .getWaitPolicy()
+            .getLifecycleStage()
+            .equals(UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED)
+        || !request.hasWaitPolicy()) {
+      throw Status.INVALID_ARGUMENT
+          .withDescription("LifeCycle stage is required")
+          .asRuntimeException();
+    }
+    if (request
+        .getWaitPolicy()
+        .getLifecycleStage()
+        .equals(UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ADMITTED)) {
+      throw Status.PERMISSION_DENIED
+          .withDescription("Admitted stage is not supported")
+          .asRuntimeException();
+    }
+    // If the workflow is in a terminal state, return the current state of the update if it
+    // completed
+    if (isTerminalState()) {
+      UpdateHandle updateHandle = getUpdate(request.getRequest().getMeta().getUpdateId());
+      if (updateHandle.getOutcome().isDone()) {
```

Comment:
Hm i was just matching the serve behavior I observed in my tests. I agree according to your diagram I should change the logic here , but the current server does not follow that behavior. I think this test shows that https://github.com/temporalio/sdk-java/pull/2153/files#diff-d14da6956e0f5f91c5901dc5dd92e9207915ca200581176fa172fd85378aa742R634. Do you think the server has a bug here?

Reply:
It's true, the server implementation diverges from the diagram in that only completed Updates are returned; I didn't actually notice that when I worked on the server change and therefore didn't call that out.

There is an explicit test case in the Server PR for this case: https://github.com/temporalio/temporal/pull/6240/files#diff-30c0e64f181e0f2ee413c578ae22f3e159748ab746563bfdeed1a9f018340611R975

The reason it is this way is that the server will abort any pending updates when the workflow completes. And that intuitively makes sense to me: otherwise a client might poll for "completed" but receiving "accepted" for the update sent to a closed workflow will keep them polling indefinitely, right? So aborting it seems like an unavoidable step.

Reply:
Thanks @stephanos, SGTM.

Reply:
>otherwise a client might poll for "completed" but receiving "accepted"

In Dans diagram though I think that would still be an error. It is when the client polls for "accepted" should it receive an error.